Tear-down address observations properly

olix0r opened the pull request:
namerd's thrift interface has no mechanism to release observations. This is
troublesome, because there is no way to obtain a reference to a *new* address
if a service is deleted and re-created.

In order to fix this, we simplify the thrift interface in the following ways:
- Do not cache Addr observations when `bind`-ing a name.
- Bind concrete names passed to the `addr` call if not cached.
- `addr` observations now take NameTree state into account.  When a concrete
  name becomes anything other than a `NameTree.Leaf`, its observation is torn
  down and removed from cache.

Fixes #237

olix0r commented:
It occurs to me that the namerd annotations should be thrift-specific since there are other namerd servers.  Let's do that separately.

adleong commented on namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala:
```diff
@@ -212,27 +279,32 @@ class ThriftNamerInterface(
+            case Activity.Pending => Var.value(Resolution.Resolved(Addr.Pending))
+            case Activity.Ok(NameTree.Leaf(bound)) if bound.id == id =>
+              bound.addr.map(Resolution.Resolved(_))
+            case _ => Var.value(Resolution.Released)
```
For clarity can we refine this pattern?  This is really `case Activity.Ok(_)`, right?  And moreover, we expect this to be triggered when it goes to `Activity.Ok(NameTree.Neg)`.  Any other value would be a violation of the assumption that the input is already a bound name, I think.

olix0r replied:
This raises an interesting point: are there situations when Activity.Ok(tree) can be satisfied where NameTree is neither Leaf nor Neg? I think it's uncommon but certainly possible given the way this currently works.  And I don't think we can expect users to know to avoid these situations.

For example, given a dtab like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
```
a bound name might be `/io.l5d.k8s/prod/http/web`.

What happens if we update the dtab so it looks like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
/io.l5d.k8s/prod => /io.l5d.k8s/pre-prod ;
```

Now `/io.l5d.k8s/prod/http/web` binds to an Alt: `/io.l5d.k8s/pre-prod/http/web | /io.l5d.k8s/prod/http/web`? (Is that right? Or does that go infinitely recursive?)

Or more simply, we may find something like:
```
/http/1.1/* => /io.l5d.k8s/prod/http ;
/io.l5d.k8s/prod/http/web => /$/empty ;
```

Perhaps we can avoid this by doing addr lookups directly against a namer instead of through via a dtab?

Or perhaps we can simply forbid dtabs that contain namer prefixes on the LHS? This seems like a generally good policy in any regard.

olix0r commented:
@adleong https://github.com/BuoyantIO/linkerd/pull/246/commits/bc3061625cb243e1601dda04196b819e61276125 alters addr resolution to skip delegation entirely so that we have a safeguard against user-provided dtabs interfering with addr binding.  This approach isn't set in stone -- let's weigh our options.

olix0r commented on namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala:
```diff
@@ -344,27 +309,75 @@ class ThriftNamerInterface(
+            case None =>
+              // could not find a namer, this addr will never bind
+              Var.value(Resolution.Resolved(Addr.Neg))
+
```
This won't work well with `/$` names like `/$/inet`.  Easy fix incoming

adleong commented:
I think I like this.  Seems like a nice separation of concerns.  Skipping delegation on the addr call makes perfect sense to me.

adleong commented on namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala:
```diff
@@ -329,50 +311,69 @@ class ThriftNamerInterface(
+        case None =>
+          Trace.recordBinary("namerd.srv/addr.cached", false)
+          val (pfx, namer) = namers.find { case (p, _) => id.startsWith(p) }.getOrElse(DefaultNamer)
+          val resolution = namer.bind(NameTree.Leaf(id.drop(pfx.size))).run.flatMap {
```
possibly factor some of this out for readability?

adleong replied:
sorry, I should have been more specific.  I meant to possibly factor out the `Activity.State => Var[Resolution]` mapping since it's long and interrupts the flow.

Not a big deal, TIOLI.

adleong commented:
:star: 👓

esbie commented:
:star: looks nice, although git's diff really makes this a difficult read :disappointed: . I'd feel more confident if there was a regression test, but don't worry if it's a pita

olix0r commented:
@esbie good point.  the thrift tests are slightly unwieldy but I'll at least give it a try.  maybe i can make them less unwieldy, even

olix0r commented:
simplified tests ThriftNamerInterface `addr` tests and added a test that properly reproduces this issue (i.e. fails as expected against master and passes on this branch).

esbie commented:
:star: :leaves: :fallen_leaf:
